x/hold/keeper/grpc_query_test.go

```diff
@@ -0,0 +1,314 @@
+		},
+		{
+			name:    "one denom on hold",
+			request: req(s.addr2.String()),
+			expResp: resp("144banana"),
+		},
+		{
+			name:    "three denoms on hold",
+			request: req(s.addr3.String()),
+			expResp: resp("89banana,55cactus,34date"),
+		},
+		{
+			name:    "error getting amount",
+			request: req(s.addr4.String()),
+			expErr: []string{
+				s.addr4.String(), "failed to read amount of dratcoin",
+				"math/big: cannot unmarshal \"dratvalue\" into a *big.Int",
+			},
+		},
+	}
+
+	for _, tc := range tests {
+		s.Run(tc.name, func() {
+			var response *hold.GetHoldsResponse
+			var err error
+			testFunc := func() {
+				response, err = s.keeper.GetHolds(s.stdlibCtx, tc.request)
+			}
+			s.Require().NotPanics(testFunc, "GetHolds")
+			s.assertErrorContents(err, tc.expErr, "GetHolds error")
+			s.Assert().Equal(tc.expResp, response, "GetHolds response")
+		})
+	}
+}
+
+func (s *TestSuite) TestKeeper_GetAllHolds() {
+	accHold := func(addr sdk.AccAddress, amount string) *hold.AccountHold {
+		return &hold.AccountHold{
+			Address: addr.String(),
+			Amount:  s.coins(amount),
+		}
+	}
+	pageResp := func(total uint64, nextKey []byte) *query.PageResponse {
+		return &query.PageResponse{
+			Total:   total,
+			NextKey: nextKey,
+		}
+	}
+	nextKey := func(addr sdk.AccAddress, denom string) []byte {
+		return keeper.CreateHoldCoinKey(addr, denom)[len(keeper.KeyPrefixHoldCoin):]
+	}
+
+	// standardSetup puts two denoms on hold for each addrs with incremental amounts.
+	// This is used unless the test has a specific setup function to use instead.
+	standardSetup := func(s *TestSuite, store sdk.KVStore) {
+		s.requireSetHoldCoinAmount(store, s.addr1, "banana", s.int(99))
+		s.requireSetHoldCoinAmount(store, s.addr1, "cherry", s.int(12))
+		s.requireSetHoldCoinAmount(store, s.addr2, "banana", s.int(100))
+		s.requireSetHoldCoinAmount(store, s.addr2, "cherry", s.int(13))
+		s.requireSetHoldCoinAmount(store, s.addr3, "banana", s.int(101))
+		s.requireSetHoldCoinAmount(store, s.addr3, "cherry", s.int(14))
+		s.requireSetHoldCoinAmount(store, s.addr4, "banana", s.int(102))
+		s.requireSetHoldCoinAmount(store, s.addr4, "cherry", s.int(15))
+		s.requireSetHoldCoinAmount(store, s.addr5, "banana", s.int(103))
+		s.requireSetHoldCoinAmount(store, s.addr5, "cherry", s.int(16))
+	}
+	standardExp := []*hold.AccountHold{
+		accHold(s.addr1, "99banana,12cherry"),
+		accHold(s.addr2, "100banana,13cherry"),
+		accHold(s.addr3, "101banana,14cherry"),
+		accHold(s.addr4, "102banana,15cherry"),
+		accHold(s.addr5, "103banana,16cherry"),
+	}
+	standardExpRev := make([]*hold.AccountHold, len(standardExp))
+	for i, val := range standardExp {
+		standardExpRev[len(standardExp)-i-1] = val
+	}
+
+	tests := []struct {
+		name        string
+		setup       func(s *TestSuite, store sdk.KVStore)
+		request     *hold.GetAllHoldsRequest
+		expHolds    []*hold.AccountHold
+		expPageResp *query.PageResponse
+		expErr      []string
+	}{
+		{
+			name:        "nil request",
```

Comment:
Ultimately, it comes down to a separation of concerns. The code in an endpoint shouldn't concern itself with the behavior of the caller. The endpoint should only concern itself with it's own behavior, and only return errors that are insurmountable.

When all parameters of the query are optional, there's no reason to require that there's a request. I.e. there's no reason to write code that returns an error if it's not actually a problem. This is common in the `GetAll` queries that only have pagination parameters. Since there's nothing in the request that's required, there's no reason to return an error when there isn't a request.

It might be possible that all code paths that end up calling the query endpoints will always provide a non-nil request, but I can't say that for certain. There might be some legitimate code paths that don't create the request object if there's nothing to put in it. For example, maybe some http get without any http query parameters ends up coming into the endpoint func with a nil request.

Reply:
Agreed, if I look at just the function then that input is fine.